hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -4326,33 +4328,59 @@ func (r *HostedClusterReconciler) reconcileKubevirtPlatformDefaultSettings(ctx c
-		// get base domain from default ingress of management cluster
-		defaultIngressOperator := manifests.IngressDefaultIngressController()
-		err := r.Client.Get(ctx, client.ObjectKeyFromObject(defaultIngressOperator), defaultIngressOperator)
-		if err != nil {
-			return err
+		// get base domain from default ingress of management/infra cluster
+		var targetClient client.Client
+		// If external infra is used, generate the infra client. Otherwise, use the client of the mgmt cluster
+		var namespace string
+		if hc.Spec.Platform.Kubevirt.Credentials != nil {
+			var err error
+			targetClient, err = generateInfraClusterClient(ctx, r.Client, *hc)
+			if err != nil {
+				return err
+			}
+			namespace = hc.Spec.Platform.Kubevirt.Credentials.InfraNamespace
+		} else {
+			targetClient = r.Client
+			namespace = hc.Namespace
 		}
 
-		// For the KubeVirt platform, the basedomain can be autogenerated using
-		// the *.apps domain of the management/infra hosting cluster
-		// This makes the guest cluster's base domain a subdomain of the
-		// hypershift infra/mgmt cluster's base domain.
-		//
-		// Example:
-		//   Infra/Mgmt cluster's DNS
-		//     Base: example.com
-		//     Cluster: mgmt-cluster.example.com
-		//     Apps:    *.apps.mgmt-cluster.example.com
-		//   KubeVirt Guest cluster's DNS
-		//     Base: apps.mgmt-cluster.example.com
-		//     Cluster: guest.apps.mgmt-cluster.example.com
-		//     Apps: *.apps.guest.apps.mgmt-cluster.example.com
-		//
-		// This is possible using OCP wildcard routes
-		hc.Spec.DNS.BaseDomain = defaultIngressOperator.Status.Domain
+		if hc.Spec.DNS.BaseDomain == "" {
+			// kubevirtInfraTempRoute is used to resolve the base domain of the infra cluster without accessing IngressController
+			kubevirtInfraTempRoute := manifests.KubevirtInfraTempRoute(namespace)
+
+			if err := targetClient.Create(ctx, kubevirtInfraTempRoute); err != nil {
+				return err
+			}
+			host := kubevirtInfraTempRoute.Spec.Host
+			if host != "" {
+				hostParts := strings.Split(host, ".")
+				baseDomain := strings.Join(hostParts[1:], ".")
+
+				// For the KubeVirt platform, the basedomain can be autogenerated using
+				// the *.apps domain of the management/infra hosting cluster
+				// This makes the guest cluster's base domain a subdomain of the
+				// hypershift infra/mgmt cluster's base domain.
+				//
+				// Example:
+				//   Infra/Mgmt cluster's DNS
+				//     Base: example.com
+				//     Cluster: mgmt-cluster.example.com
+				//     Apps:    *.apps.mgmt-cluster.example.com
+				//   KubeVirt Guest cluster's DNS
+				//     Base: apps.mgmt-cluster.example.com
+				//     Cluster: guest.apps.mgmt-cluster.example.com
+				//     Apps: *.apps.guest.apps.mgmt-cluster.example.com
+				//
+				// This is possible using OCP wildcard routes
+				hc.Spec.DNS.BaseDomain = baseDomain
+
+				if err := targetClient.Delete(ctx, kubevirtInfraTempRoute); err != nil {
+					return err
+				}
+			}
```

Comment:
The conditional logic `if host != ""` makes this code look a little problematic. I know the logic works at the moment, but the way it's structured put's us slightly at risk. Here's why

If somehow the route's `host == ""`, then the temp route will not be deleted, the BaseDomain will remain unset, this function will silently move on. However, on the next reconcile, we'd wind up right back here again because BaseDomain is still empty. Except this time the `Create` will fail because the route already exists, and we'll continue to permanently fail the reconcile from that point because the logic that cleans up the temp route will never execute. The error will make it look like we're failing to "create" something, when in reality the error should have been that we couldn't detect a host value.

I think two simple changes would help here.

1. change the Create() logic for the temp route to use the CreateOrUpdate() logic instead. I know this involves passing in that createOrUpdate function pointer through a few layers of functions, but it shouldn't be too big of a deal.
2. return an error if `host == ""`something like ... "unable to autodetect kubevirt platform BaseDomain, temporary route is not giving a host value"

With those two changes, we'll at least be bubbling up the right error if this (route with empty host) does somehow occur in the future. And it also gives the error a chance to resolve if the route get's the `host` value lazily applied later on for some reason.

Reply:
done, now using the `createOrUpdate` function for the temp route.

Reply:
done, updated the function to use `CreateOrUpdate` from `upsert` instead of the regular controller runtime `Create`